### PR TITLE
Add method to download artifacts

### PR DESCRIPTION
**Pull request by ostefano**
Signed-off-by: Stefano Ortolani <ostefano@gmail.com>

### CODE REVIEW

File: examples/submit_samples.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    submission_start_ts = analysis_client.get_api_utc_timestamp()
+    submissions = []
+    task_to_source = {}
+    for file_path in file_paths:
+        with open(file_path, "rb") as f:
+            ret = analysis_client.submit_file(f.read(), bypass_cache=args.bypass_cache)
+            submissions.append(ret)
+            task_to_source[ret["task_uuid"]] = file_path
+    for file_hash in file_hashes:
+        file_data = download_from_vt(vt_client, file_hash)
```
Comment (Phat3) [inline examples/submit_samples.py:278]: Here `vt_client` might be None

File: examples/submit_samples.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            if ret["more_results_available"]:
+                continue
+            if not pending_submissions:
+                break
+        time.sleep(_get_timeout_or_raise())
+    print(f"Done waiting for completion of {len(submissions)} submissions")
+
+
+def is_likely_binary(file_path: str) -> bool:
+    """Return whether the file is likely to be binary."""
```
Comment (Phat3) [inline examples/submit_samples.py:99]: Missing docstring for parameter

File: examples/submit_samples.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        with open(file_path, "r") as f:
+            for _ in f:
+                break
+            return False
+    except UnicodeDecodeError:
+        return True
+
+
+def detect_input_type(input_bit) -> str:
+    """Detect the input type."""
```
Comment (Phat3) [inline examples/submit_samples.py:110]: Missing docstring for parameter

File: examples/submit_samples.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    try:
+        buffer = io.BytesIO()
+        client.download_file(file_hash, buffer)
+        buffer.seek(0, 0)
+        return buffer.read()
+    except (IOError, vt.APIError) as e:
+        raise ValueError(str(e)) from e
+
+
+def is_hash(input_bit: str) -> bool:
```
Comment (Phat3) [inline examples/submit_samples.py:143]: Missing docstring for parameter

File: examples/submit_samples.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        return False
+    else:
+        try:
+            int(input_bit, 16)
+            return True
+        except ValueError:
+            return False
+
+
+def get_file_hashes(file_path: str, limit: Optional[int] = None) -> List[str]:
```
Comment (Phat3) [inline examples/submit_samples.py:156]: Missing docstring for parameters

File: examples/submit_samples.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    random.Random(4).shuffle(file_hashes)
+    return file_hashes
+
+
+def get_file_paths(
+    dir_name: str,
+    extension: Optional[str] = None,
+    limit: Optional[int] = None,
+) -> List[str]:
+    """Get all file paths inside a directory."""
```
Comment (Phat3) [inline examples/submit_samples.py:178]: Missing docstring for parameters

File: src/tau_clients/__init__.py [hunk 7]
```diff
@@ -36,12 +44,52 @@
...
+REPORT_TYPE_SANDBOX = "ll-int-win"
+
+# Datetime formats
+DATETIME_FMT = "%Y-%m-%d %H:%M:%S"
+DATETIME_MSEC_FMT = DATETIME_FMT + ".%f"
+DATE_FMT = "%Y-%m-%d"
+
+
+def is_likely_task_uuid(hash_like: str) -> bool:
+    """Heuristic to check whether a string is likely a task uuid."""
```
Comment (Phat3) [inline src/tau_clients/__init__.py:65]: Missing docstring for parameter

File: src/tau_clients/__init__.py [hunk 8]
```diff
@@ -74,12 +122,12 @@ def get_task_link(
     :raises ValueError: if not enough parameters have been provided
```
Comment (Phat3) [inline src/tau_clients/__init__.py:122]: The method does not raise this exception anymore

File: examples/submit_samples.py (hunk 1 above)
Comment (ostefano) [reply examples/submit_samples.py:278]: Not really, because if before was None file_hashes is empty here

File: examples/submit_samples.py (hunk 6 above)
Comment (ostefano) [reply examples/submit_samples.py:178]: ACK

File: examples/submit_samples.py (hunk 5 above)
Comment (ostefano) [reply examples/submit_samples.py:156]: ACK

File: examples/submit_samples.py (hunk 4 above)
Comment (ostefano) [reply examples/submit_samples.py:143]: ACK

File: examples/submit_samples.py (hunk 3 above)
Comment (ostefano) [reply examples/submit_samples.py:110]: ACK

File: examples/submit_samples.py (hunk 2 above)
Comment (ostefano) [reply examples/submit_samples.py:99]: ACK

File: examples/download_artifacts.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            for artifact_type in artifact_types:
+                nb_count = len([x for x in results if x["artifact_type"] == artifact_type])
+                print(f"[{file_hash}][{task_uuid}] Found {nb_count} {artifact_type}(s)")
+            for result in results:
+                hash_to_artifact_names[file_hash].append(result)
+
+    # Download all artifacts
+    for file_hash, artifact_names in hash_to_artifact_names.items():
+        for artifact_name in artifact_names:
+            file_name = "{file_hash}.{task_uuid}.{report_uuid}.{artifact_name}".format(
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:163]: ACK

File: examples/download_artifacts.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        hash_to_tasks = collections.defaultdict(set)
+        for file_hash in args.file_inputs:
+            ret = analysis_client.query_file_hash(file_hash)
+            for task in ret.get("tasks", []):
+                hash_to_tasks[file_hash].add(task["task_uuid"])
+            if file_hash in hash_to_tasks:
+                print(f"File {file_hash} found")
+            else:
+                print(f"File {file_hash} NOT found")
+    else:
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:138]: ACK

File: examples/download_artifacts.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    report_types = [] if args.disable_sandbox_filter else [tau_clients.REPORT_TYPE_SANDBOX]
+    report_types_str = ",".join(report_types) or "all"
+
+    # Decode input type
+    input_type = args.input_type
+    if not input_type:
+        if all(tau_clients.is_likely_task_uuid(x) for x in args.file_input):
+            print("Input will be treated as task uuids (use '-u' to force input type)")
+            input_type = INPUT_TYPE_TASK_UUID
+        else:
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:110]: ACK fixed

File: examples/download_artifacts.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    conf.read(args.config_file)
+
+    # Load the analysis client
+    analysis_client = nsx_defender.MultiAnalysisClient.from_conf(conf, "analysis")
+
+    # Parse the input
+    artifact_types = METADATA_TYPES if "all" in args.artifact_types else args.artifact_types
+    artifact_types_str = ",".join(artifact_types)
+    report_types = [] if args.disable_sandbox_filter else [tau_clients.REPORT_TYPE_SANDBOX]
+    report_types_str = ",".join(report_types) or "all"
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:102]: ACK

File: examples/download_artifacts.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    )
+    parser.add_argument(
+        "file_inputs",
+        metavar="file_inputs",
+        nargs="+",
+        help="file hashes or task uuids",
+    )
+    args = parser.parse_args()
+    if not os.path.isfile(args.config_file):
+        raise ValueError(f"No configuration file at file path '{args.config_file}'")
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:91]: ACK

File: examples/download_artifacts.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+    tau_clients.METADATA_TYPE_PCAP,
+    tau_clients.METADATA_TYPE_PROCESS_SNAPSHOT,
+    tau_clients.METADATA_TYPE_YARA_STRINGS,
+    tau_clients.METADATA_TYPE_CODEHASH,
+    tau_clients.METADATA_TYPE_SCREENSHOT,
+    tau_clients.METADATA_TYPE_SFC,
+]
+
+
+def config_file(file_path: str) -> str:
```
[the remark this answers is not in the document]
Comment (ostefano) [reply examples/download_artifacts.py:30]: ACK

File: src/tau_clients/__init__.py (hunk 7 above)
Comment (ostefano) [reply src/tau_clients/__init__.py:65]: ACK

**Pull request merged**